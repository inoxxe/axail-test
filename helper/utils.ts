export function findCheapestPrice(
  n: number,
  flights: number[][],
  src: number,
  dst: number,
  k: number,
): number {
  let prices = new Array(n).fill(Infinity);
  prices[src] = 0;
  for (let stopI = 0; stopI <= k; stopI++) {
    const currPrices = Array.from(prices);
    for (const [flightFrom, flightTo, flightPrice] of flights) {
      const newPrice = prices[flightFrom] + flightPrice;
      if (newPrice < currPrices[flightTo]) {
        currPrices[flightTo] = newPrice;
      }
    }
    prices = currPrices;
  }
  return prices[dst] === Infinity ? -1 : prices[dst];
}
