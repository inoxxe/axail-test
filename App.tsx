import React, {useState} from 'react';
import {
  Text,
  StyleSheet,
  TouchableOpacity,
  ScrollView,
  SafeAreaView,
  Alert,
} from 'react-native';
import {isCancel, isInProgress, pick} from 'react-native-document-picker';
import RNFS from 'react-native-fs';
import {findCheapestPrice} from './helper/utils';

type Input = {
  n: number;
  edges: number[][];
  start: number;
  end: number;
  expected: number;
};

const App = () => {
  const [input, setInput] = useState<Input | null>(null);

  const handleError = (err: unknown) => {
    if (isCancel(err)) {
      Alert.alert('Error', 'User cancelled the picker');
      // User cancelled the picker, exit any dialogs or menus and move on
    } else if (isInProgress(err)) {
      Alert.alert(
        'Error',
        'multiple pickers were opened, only the last will be considered',
      );
    } else {
      throw err;
    }
  };

  const pickDocument = async () => {
    try {
      const res = await pick({
        allowMultiSelection: false,
        type: ['application/json'],
      });
      readDocumentData(res[0].uri);
    } catch (err) {
      handleError(err);
    }
  };

  const readDocumentData = async (uri: string) => {
    RNFS.readFile(uri, 'ascii')
      .then(res => {
        const d = JSON.parse(res);
        if (
          !d.hasOwnProperty('n') ||
          !d.hasOwnProperty('edges') ||
          !d.hasOwnProperty('start') ||
          !d.hasOwnProperty('end') ||
          !d.hasOwnProperty('expected')
        ) {
          Alert.alert(
            'Error',
            'The selected file does not meet the required format',
          );
          return;
        }
        setInput(d);
      })
      .catch(err => {
        console.log(err.message, err.code);
      });
  };

  return (
    <SafeAreaView style={styles.container}>
      <ScrollView contentContainerStyle={styles.scrollContainer}>
        <Text style={styles.title}>Question Answer App</Text>

        <TouchableOpacity
          testID="pick-button"
          style={styles.button}
          onPress={pickDocument}>
          <Text style={styles.buttonText}>Upload JSON File</Text>
        </TouchableOpacity>

        {input && (
          <Text>
            Your cheapest price from {input?.start} to {input?.end} is{' '}
            {findCheapestPrice(
              input?.n,
              input?.edges,
              input?.start,
              input?.end,
              input?.expected,
            )}
          </Text>
        )}
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f5f5f5',
  },
  scrollContainer: {
    padding: 20,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    marginBottom: 20,
    textAlign: 'center',
  },
  button: {
    backgroundColor: '#007AFF',
    padding: 15,
    borderRadius: 8,
    marginBottom: 20,
  },
  buttonText: {
    color: '#fff',
    textAlign: 'center',
    fontSize: 16,
    fontWeight: '600',
  },
  card: {
    backgroundColor: '#fff',
    padding: 15,
    borderRadius: 8,
    marginBottom: 15,
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
  },
  label: {
    fontSize: 16,
    fontWeight: '600',
    marginBottom: 5,
    color: '#333',
  },
  content: {
    fontSize: 16,
    color: '#666',
  },
  error: {
    color: '#ff3b30',
    textAlign: 'center',
    marginTop: 10,
  },
});

export default App;
