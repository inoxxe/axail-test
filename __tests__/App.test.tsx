import React from 'react';
import {render} from '@testing-library/react-native';
import App from '../App';
import {findCheapestPrice} from '../helper/utils';

describe('App Component', () => {
  it('renders correctly', () => {
    const {getByText} = render(<App />);
    expect(getByText('Question Answer App')).toBeTruthy();
    expect(getByText('Upload JSON File')).toBeTruthy();
  });

  it('should return the correct cheapest price', () => {
    const flights = [
      [0, 1, 100],
      [1, 2, 100],
      [0, 2, 500],
    ];
    const input = {
      n: 3,
      edges: flights,
      start: 0,
      end: 2,
      expected: 1,
    };

    const result = findCheapestPrice(
      input.n,
      flights,
      input.start,
      input.end,
      input.expected,
    );

    expect(result).toBe(200);
  });
});
