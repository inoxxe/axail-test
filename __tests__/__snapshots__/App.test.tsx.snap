// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#f5f5f5",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "padding": 20,
      }
    }
  >
    <View>
      <Text
        style={
          {
            "fontSize": 24,
            "fontWeight": "bold",
            "marginBottom": 20,
            "textAlign": "center",
          }
        }
      >
        Question Answer App
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#007AFF",
            "borderRadius": 8,
            "marginBottom": 20,
            "opacity": 1,
            "padding": 15,
          }
        }
      >
        <Text
          style={
            {
              "color": "#fff",
              "fontSize": 16,
              "fontWeight": "600",
              "textAlign": "center",
            }
          }
        >
          Upload JSON File
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
